### PR TITLE
refactor: Restructure Avo::BaseResource for user customization

### LINKED ISSUE #2959: Create a blank `Avo::BaseResource`
We should move the contents of `Avo::BaseResource` to something else and leave a blank `Avo::BaseResource` instead.
This will enable our users to override it in their own apps so they can add their own custom methods more easily.

This should be a best practice for us to use in more places.

`Avo::Resources::Post` inherits from `Avo::BaseResource` which inherits from `Avo::Resources::Base` (or some other name)

The user can add their own `Avo::BaseResource` and add custom methods that all of their resources will inherit.

```ruby
class Avo::BaseResource < Avo::Resources::Base
  # make all number fields cast their values `to_f`
  def field(id, **args, &)
    if args[:as] == :number
      args[:format_using] = -> { value.to_f }
    end

    super id, **args, &
  end
end
```

### PR DESCRIPTION
**Pull request by Yash-Singh-Pathania**
# Description

Fixes #2959 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the [documentation](https://github.com/avo-hq/docs.avohq.io/pull/247)
- [ ] I have added tests that prove my fix is effective or that my feature works

## Manual review steps

1. Ran local test cases 
2. Did dev testing via rails console.

### CODE REVIEW

File: lib/avo/resources/base.rb [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+          record: record
+        }
+      end
+
+      def entity_loader(entity)
+        instance_variable_get("@#{entity.to_s.pluralize}_loader")
+      end
+    end
+  end
+end
```
Comment (Paul-Bob) [inline lib/avo/resources/base.rb:606]:
```suggestion
end

```

**Review by Paul-Bob — COMMENTED** [review]
Looking good, let's please add some documentation about this something like we have on controllers https://docs.avohq.io/3.0/resources.html#extending-avo-resourcescontroller

File: lib/avo/resources/base.rb (hunk 1 above)
Comment (Yash-Singh-Pathania) [reply lib/avo/resources/base.rb]: Changes made added the empty line sorry for that , also @Paul-Bob  would be totally down to write the docs could you just guide me a the repository from what I see is different , so do I raise a seprate pr and link it to this pr ?

But yeah totally down to writing it . 
lmk
Thanks

Comment (Paul-Bob) [reply lib/avo/resources/base.rb]: Yes, this is the docs repo https://github.com/avo-hq/docs.avohq.io

You can add a PR there and link it on this PR description, something like:

```
...
Checklist
...
- [x] I have made corresponding changes to the [documentation](https://github.com/avo-hq/avodocs)
  - Link to docs here <--
...
```

Let me know if you need any help setting up the docs repo

Thank you!

**Comment by Paul-Bob** [comment]
BTW let's ignore the lint on this PR, let's keep the same code, we can lint it later

**Comment by Yash-Singh-Pathania** [comment]
@Paul-Bob added empty lines at the end without making other changes, regarding lining. Thanks for the help, but I got the docs set up and am adding a pull request for the same. Thanks.

**Comment by Yash-Singh-Pathania** [comment]
Find document changes [here]( https://github.com/avo-hq/docs.avohq.io/pull/247)
@Paul-Bob

**Comment by Yash-Singh-Pathania** [comment]
@Paul-Bob  is there anything else needed on this pr ? 
Or are we good to go

**Comment by Paul-Bob** [comment]
Hey @Yash-Singh-Pathania I'll do a more careful review tomorrow, looking good! Thank you!

**Review by Paul-Bob — APPROVED** [review]
> @Paul-Bob  is there anything else needed on this pr ? 
Or are we good to go 

We are good to go!

We appreciate your contribution

**Pull request merged**